zwe init typescript edition

1000TurquoisePogs opened the pull request:
- [ ] Tests for the changes have been added (for bug fixes / features)
- [x] DCO signoffs have been added to all commits, including this PR

#### PR type
- [x] Feature <!-- non-breaking change which adds functionality -->

#### Changes proposed in this PR
<!-- Please describe your Pull Request. -->
- This adds all zwe init commands into configmgr editions except for zwe init certificate because it requires a whole lot more code.

#### Does this PR introduce a breaking change?
- [x] No


#### Does this PR do something the person installing Zowe should know about?
This is quite a bit of code and the tmp files it writes out are very important to check for correctness, so do not merge this yet, it is still being tested and then test cases will be stated.

1000TurquoisePogs committed:
Very WIP on init mvs and init vsam

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

1000TurquoisePogs committed:
Check-in for still very wip code

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

1000TurquoisePogs committed:
Working on stc index

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

1000TurquoisePogs committed:
Fixes for compile

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

1000TurquoisePogs committed:
WIP on cleaning up all the grep and sed stuff

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

1000TurquoisePogs committed:
Compile fixes

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

1000TurquoisePogs committed:
Merge branch 'v2.x/master' into feature/v2/zwe-init-ts

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

1000TurquoisePogs committed:
Merge branch 'v2.x/staging' into feature/v2/zwe-init-ts

1000TurquoisePogs committed:
Fixes upon testing

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

1000TurquoisePogs committed:
More fixes on testing

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

1000TurquoisePogs committed:
Fix path for zwe

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

1000TurquoisePogs committed:
Fixes for undefined params

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

1000TurquoisePogs committed:
Merge branch 'feature/v2/zwe-init-ts' of github.com:zowe/zowe-install-packaging into feature/v2/zwe-init-ts

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

1000TurquoisePogs committed:
Fix flipped boolean on allowoverwrite

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

1000TurquoisePogs committed:
Flip log codepage on zos

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

1000TurquoisePogs committed:
Resort to using shell for logging at the moment

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

1000TurquoisePogs committed:
Add new type

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

1000TurquoisePogs committed:
Respect zwe bad RCs. Attempt to fix log file output with new command, and fix zosmf trust behavior

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

1000TurquoisePogs committed:
Merge branch 'v2.x/staging' into feature/v2/zwe-init-ts

1000TurquoisePogs committed:
Fix build

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

1000TurquoisePogs commented on bin/commands/init/index.ts:
```diff
@@ -0,0 +1,116 @@
+    initApfAuth.execute();
+    initSecurity.execute();
+  }
+//  initCertificate.execute();
```
reintroduce this

1000TurquoisePogs commented on bin/commands/init/index.ts:
```diff
@@ -0,0 +1,116 @@
+import * as initVsam from './vsam/index';
+import * as initApfAuth from './apfauth/index';
+import * as initSecurity from './security/index';
+//import * as initCertificate from './certificate/index';
```
reintroduce this

1000TurquoisePogs commented on bin/commands/init/mvs/index.ts:
```diff
@@ -0,0 +1,119 @@
+    if (!ds) {
+      // authLoadlib can be empty
+      if (datasetDef.configKey == 'authLoadlib') {
+        skip=true;
```
need to set skip to false in else case or else you'd be skipping all the rest on the next loop

1000TurquoisePogs replied:
i read it wrong, its fine as it is.

1000TurquoisePogs commented on bin/commands/init/vsam/index.ts:
```diff
@@ -0,0 +1,173 @@
+      common.printErrorAndExit(`Error ZWEL0163E: Job ${jobname}(${jobid}) ends with code ${jobcccode} (${jobcctext}).`, undefined, 163);
+    }
+  } else {
+    common.printErrorAndExit(`Error ZWEL0999E: Waiting for job failed, jobname parsing failed.`);
```
make an error code, or remove!

1000TurquoisePogs commented on bin/commands/init/security/index.ts:
```diff
@@ -0,0 +1,173 @@
+                          .replace(/\n\/\/\s+SET\s+ZISSTC=.*\n/, `\n//         SET  ZISSTC=${securityStcsZis}\n`)
+                          .replace(/\n\/\/\s+SET\s+AUXSTC=.*\n/, `\n//         SET  AUXSTC=${securityStcsAux}\n`)
+                          .replace(/\n\/\/\s+SET\s+HLQ=.*\n/, `\n//         SET  HLQ=${prefix}\n`)
+                          .replace(/\n\/\/\s+SET\s+SYSPROG=.*\n/, `\n//         SET  SYSPROG=${securityGroupsSysprog}\n`));
```
is this working?

Martin-Zeithaml replied:
Seems to work fine on our system.

I was able to broke this only by manual editing of source `SAMP(ZWESECUR)`, for example by adding label to `SET` statement: `//PRODUCT    SET  PRODUCT=RACF` (but this is too obvious, that it will not work).

1000TurquoisePogs commented on bin/commands/init/vsam/index.ts:
```diff
@@ -0,0 +1,173 @@
+    common.printErrorAndExit(`Error ZWEL0162E: Failed to find job ${jobid} result.`, undefined, 162);
+  }
+  const sections = jobstate.out.split(',');
+  if (sections.length >= 3) {
```
>=4 actually

1000TurquoisePogs commented on bin/commands/init/certificate/index.ts:
```diff
@@ -0,0 +1,523 @@
+  CERT_PARMS.pkcs12_name_lc=CERT_PARMS.pkcs12_name.toLowerCase();
+  CERT_PARMS.pkcs12_caAlias_lc=CERT_PARMS.pkcs12_caAlias.toLowerCase();
+  // what PEM format CAs we should tell Zowe to use
+  delete CERT_PARMS.yaml_pem_cas;
```
set to empty string????

1000TurquoisePogs replied:
fine due to undefined wrapper logic below

1000TurquoisePogs commented on bin/libs/string.ts:
```diff
@@ -252,6 +252,10 @@ export function paddingLeft(str: string, pad: string): string {
 }
 
+export function stripUndefined(input: string): string {
+  return input.replace(/undefined/g, '');
```
surround with quotes?

1000TurquoisePogs replied:
there's quotes on the certificates.ts file already, its fine as is

1000TurquoisePogs committed:
Fixes from review

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

1000TurquoisePogs committed:
Fix and simplify the waitforjob logic

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

1000TurquoisePogs committed:
Fix that zwe was missing parms when calling it within itself during init certificate

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

1000TurquoisePogs committed:
Merge branch 'v2.x/staging' into feature/v2/zwe-init-ts

Martin-Zeithaml commented on bin/commands/init/certificate/index.ts:
```diff
@@ -0,0 +1,531 @@
+  if (!certType) {
+    common.printErrorAndExit(`Error ZWEL0157E: Certificate type (zowe.setup.certificate.type) is not defined in Zowe YAML configuration file.`, undefined, 157);
+  }
+  if (certType != "PKCS12" && certType != "JCERACFKS") {
```
Is this up to date? 
https://github.com/zowe/zowe-install-packaging/blob/610827598b1c3deeb3a8467ab2745c1a0b04edef/example-zowe.yaml#L100

1000TurquoisePogs replied:
no, this is months old and needs rebasing.

Martin-Zeithaml commented on bin/commands/init/certificate/index.ts:
```diff
@@ -0,0 +1,531 @@
+    // should we clean up before creating new
+    if (std.getenv('ZWE_CLI_PARAMETER_ALLOW_OVERWRITE') == "true") {
+      // warning
+      common.printMessage(`Warning ZWEL0300W: Keyring "safkeyring:///${CERT_PARMS.keyring_owner}/${CERT_PARMS.keyring_name}" will be overwritten during configuration.`);
```
Will print `safkeyring:///USERID...`, (same as in shell version). It is already a little bit confusing, that `safkeyring` is used with two or four slashes and now even with three 😄 
ZWEL0300W not documented in `zwe init certificate` section

1000TurquoisePogs replied:
3 slashes is certainly a bug to fix.
4 slashes is needed for apiml unless that bug got fixed. other servers tolerate 2/4
`.errors` files document which errors exist in which command
https://github.com/zowe/zowe-install-packaging/blob/eefffa903fc0d2730ce43f7a89e61f4e2971c60e/bin/commands/init/certificate/.errors

This doesnt contain 300w, but should. If you make that edit, then the docs site has some automation that puts it into the documentation.

Martin-Zeithaml commented on bin/commands/init/apfauth/index.ts:
Success for `zwe init apfauth -c zowe.yaml`

Where zowe.yaml is:
```
zowe:
  setup:
    dataset:
      prefix: USERID.ZOWE.ZWEINIT
      parmlib: ${{ zowe.setup.dataset.prefix }}.PARMLIB
      jcllib: ${{ zowe.setup.dataset.prefix }}.JCLLIB
      authLoadlib:  ${{ zowe.setup.dataset.prefix }}.TEST1
      authPluginLib:  ${{ zowe.setup.dataset.prefix }}.TEST2
```
👍

Martin-Zeithaml commented on bin/commands/init/mvs/index.ts:
Success for `./zwe init mvs -c ./zweMVS.yaml --allow-overwrite` and `./zwe init mvs -c ./zweMVS.yaml`

Where zweMVS.yaml is:

```
zowe:
  setup:
    dataset:
      prefix: USERID.ZOWE.ZWEINIT
      parmlib: ${{ zowe.setup.dataset.prefix }}.PARMLIB
      jcllib: ${{ zowe.setup.dataset.prefix }}.JCLLIB
      authLoadlib:  ${{ zowe.setup.dataset.prefix }}.TEST1
      authPluginLib:  ${{ zowe.setup.dataset.prefix }}.TEST2
  useConfigmgr: true
  runtimeDirectory: ../
```

I was executing this from `bin` directory, that's why the `runtimeDirectory` is `../` and it worked.

1000TurquoisePogs replied:
relative paths being allowed in runtime directory is scary to me, but if we allow it then ok.

Martin-Zeithaml replied:
When I realize, that I need (for the testing purposes ) to go only one level up in the directory hierarchy, I was curious if the relative path will work 😄

Martin-Zeithaml replied:
```
$ pwd
/z/user123/git/qa.zweInit/mvs/ttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt/ttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt/ttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt
$ ls -al
total 32
drwxr-xr-x   2 USER123  OMVSGRP     8192 Jun 20 08:17 .
drwxr-xr-x   3 USER123  OMVSGRP     8192 Jun 20 08:15 ..
lrwxrwxrwx   1 USER123  OMVSGRP       23 Jun 20 08:17 runtime -> /z/user123/git/zweInit/
```
MVS_ER.yaml:
```
zowe:
  setup:
    dataset:
      prefix: USER123.ZOWE.ZWEINIT
      parmlib: ${{ zowe.setup.dataset.prefix }}.PARMLIB
      jcllib: ${{ zowe.setup.dataset.prefix }}.JCLLIB
      authLoadlib:  ${{ zowe.setup.dataset.prefix }}.TEST1
      authPluginLib:  ${{ zowe.setup.dataset.prefix }}.TEST2
  useConfigmgr: true
  runtimeDirectory: /z/user123/git/qa.zweInit/mvs/ttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt/ttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt/ttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttttt/runtime
```
When executed:
```
$ /z/user123/git/zweInit/bin/zwe init mvs -c MVS_ER.yaml
Temporary directory '/tmp/.zweenv-5992' created.
Zowe will remove it on success, but if zwe exits with a non-zero code manual cleanup would be needed.
```
It has this strange behavior, no errors, no warnings, nothing was done. The zwe config validate or get same behavior.

Seems to be a problem of handling long strings. After few attempts there is mysterious limit of 432 chars for `runtimeDirectory` (the z/os uss limit is in total 1023 chars and 255 for file). I didn't test it yet, but I am expecting the same issue with other directories or similar.

The `runtimeDirectory` string was in one line or multiple lines with the same strange result.

_Note:_
Is the setting in MVS_ER.yaml stupid?  **YES!**
But is it all valid and legal? **Also YES!**

Martin-Zeithaml commented on bin/commands/init/security/index.ts:
```
./zwe init security -c ./zweSecur.yaml --security-dry-run

zowe:
  setup:
    dataset:
      prefix: ZEIMA02.ZOWE.ZWEINIT
      jcllib: ${{ zowe.setup.dataset.prefix }}.JCLLIB
    security:
      product: ACF2 # TSS | RACF | ACF2
      groups:
        admin: ZW3ADMIN_LONGNAME
        stc: ZW3ADMIN
        sysProg: ZW3ADMIN
      users:
        zowe: ZW3SVUSR
        zis: ZW3SIUSR
      stcs:
        zowe: ZW3SLSTC_LONG_NAME
        zis: ZW3SISTC
        aux: '1234567890'
  useConfigmgr: true
  runtimeDirectory: ../
```

```
...
//         EXPORT SYMLIST=*                                          
//*                                                                  
//         SET  PRODUCT=ACF2                                         
//*                     12345678                                     
//         SET  ADMINGRP=ZW3ADMIN_LONGNAME                           
//         SET  STCGRP=ZW3ADMIN                                      
//         SET  ZOWEUSER=ZW3SVUSR                                    
//         SET  ZISUSER=ZW3SIUSR                                     
//         SET  ZOWESTC=ZW3SLSTC_LONG_NAME                           
//         SET ZLNCHSTC=ZWESLSTC     * Zowe started task name for HA 
//         SET  ZISSTC=ZW3SISTC                                      
//         SET  AUXSTC=1234567890                                    
//         SET  HLQ=ZEIMA02.ZOWE.ZWEINIT                             
//         SET  SYSPROG=ZW3ADMIN                                     
//*                     12345678   
...                                  
```

1) But if the `SZWESAMP` is empty/does not exist, the command is stuck here and must be terminated:
```
>> Run Zowe security configurations

Modify ZWESECUR
cat: //'ZEIMA02.ZOWE.ZWEINIT.SZWESAMP(ZWESECUR)': EDC5049I The specified file name could not be located. (errno2=0xC00B0641)
CEE5206S The signal SIGINT was received.
```
If the JCLLIB is used (e.g. in ISPF), the process ends with the correct error message.

2) Not caused by this PR, but there might be pretty much anything as admin, stc, user... E.g. AUXSTC is numeric only.

3) Garbage in the command output
```
>> Run Zowe security configurations

*** Modify ZWESECUR ***

��/��ʀ�/�%������ǀ���>?���- ZEIMA02.ZOWE.ZWEINIT.JCLLIB(ZW141317) is prepared

Dry-run mode, security setup is NOT performed on the system.
Please submit ZEIMA02.ZOWE.ZWEINIT.JCLLIB(ZW141317) manually.
Temporary directory /tmp/.zweenv-6640 removed successfully.
```
Seems to be done by `zosfs.ensureFileEncoding(tmpfile, "SPDX-License-Identifier");`

Martin-Zeithaml commented on bin/commands/init/stc/index.ts:
```diff
@@ -0,0 +1,273 @@
+
+      const tmpFileContent = sistcContent.out.replace('ZWES.SISSAMP', parmlib)
+        .split('\n').map(line => line.startsWith('//STEPLIB') ? `//STEPLIB  DD   DSNAME='${authLoadlib}',DISP=SHR\n`
+                                                              + `//         DD   DSNAME='${authPluginLib}',DISP=SHR'`
```
Extra apostrophe at the end of line 4:
```
//ZWESIS01 EXEC PGM=ZWESIS01,REGION=&RGN,                       
//         PARM='NAME=&NAME,MEM=&MEM'                           
//STEPLIB  DD   DSNAME='ZEIMA02.ZOWE.ZWEINIT.TEST1',DISP=SHR    
//         DD   DSNAME='ZEIMA02.ZOWE.ZWEINIT.TEST2',DISP=SHR'   
//PARMLIB  DD   DSNAME=ZEIMA02.ZOWE.ZWEINIT.PARMLIB,DISP=SHR    
//SYSPRINT DD   SYSOUT=*                                        
```

Martin-Zeithaml commented on bin/commands/init/stc/index.ts:
```diff
@@ -0,0 +1,273 @@
+      
+      const tmpFileContent = sastcContent.out.split('\n')
+        .map(line => line.startsWith('//STEPLIB') ? `//STEPLIB  DD   DSNAME='${authLoadlib}',DISP=SHR\n`
+                                                  + `//         DD   DSNAME='${authPluginLib}',DISP=SHR'`
```
Extra apostrophe at the end of line

Martin-Zeithaml commented on bin/commands/init/stc/index.ts:
```diff
@@ -0,0 +1,273 @@
+      common.printTrace(`  * Output:`);
+      common.printTrace(stringlib.paddingLeft(slstcContent.out, "    "));
+
+      const tmpFileContent = slstcContent.out.replace('//STEPLIB  DD   DSNAME=&SYSUID..LOADLIB,DISP=SHR', '//STEPLIB  DD   DSNAME=${authLoadlib},DISP=SHR')
```
Does not work and produce variable name in STEPLIB:
`//STEPLIB  DD   DSNAME=${authLoadlib},DISP=SHR`

Backtick or plus:
```
const tmpFileContent = slstcContent.out.replace('//STEPLIB  DD   DSNAME=&SYSUID..LOADLIB,DISP=SHR', '//STEPLIB  DD   DSNAME='+authLoadlib+',DISP=SHR')
or 
const tmpFileContent = slstcContent.out.replace('//STEPLIB  DD   DSNAME=&SYSUID..LOADLIB,DISP=SHR', `//STEPLIB  DD   DSNAME=${authLoadlib},DISP=SHR`)
```

Martin-Zeithaml commented on bin/commands/init/stc/index.ts:
`zweSTC.yaml` is
```
zowe:
  setup:
    dataset:
      prefix: ZEIMA02.ZOWE.ZWEINIT
      parmlib: ${{ zowe.setup.dataset.prefix }}.PARMLIB
      proclib: ${{ zowe.setup.dataset.prefix }}.PROCLIB
      jcllib: ${{ zowe.setup.dataset.prefix }}.JCLLIB
      authLoadlib:  ${{ zowe.setup.dataset.prefix }}.TEST1
      authPluginLib:  ${{ zowe.setup.dataset.prefix }}.TEST2
    security:
      stcs:
        zowe: MZ1
        zis: '2'
        aux: MZ23_AUX
  useConfigmgr: true
  runtimeDirectory: ../
```
### Find by accident - RESOLVED
There is only one member in SZWESAMP: `ZEIMA02.ZOWE.ZWEINIT.SZWESAMP(ZWESECUR)`
`zwe init stc -c zweSTC.yaml`

The following error message does not make sense to me:
```
>> Install Zowe main started task

ERROR: Error ZWEL0143E: ZEIMA02.ZOWE.ZWEINIT.SZWESAMP(ZWESLSTC) already exists. This data set member will be overwritten during configuration.
```

### SZWESAMP OK
`ZEIMA02.ZOWE.ZWEINIT.SZWESAMP` is complete (all expected members are there)

`zwe init stc -c zweSTC.yaml`

```
>> Install Zowe main started task

Modify ZWESLSTC and save as ZEIMA02.ZOWE.ZWEINIT.JCLLIB(MZ1)
  > Check if either TMPDIR or TMP points to writable directory, else try '/tmp' directory
- Copy ZEIMA02.ZOWE.ZWEINIT.SZWESAMP(ZWESLSTC) to /tmp/zwe-init-stc-1549
  * Succeeded
��/��ʀ�/�%������ǀ���>?���- cp  -v "/tmp/zwe-init-stc-1549" "//'ZEIMA02.ZOWE.ZWEINIT.JCLLIB(MZ1)'"
  * Succeeded
- ZEIMA02.ZOWE.ZWEINIT.JCLLIB(MZ1) is prepared
Modify ZWESISTC and save as ZEIMA02.ZOWE.ZWEINIT.JCLLIB(2)
  > Check if either TMPDIR or TMP points to writable directory, else try '/tmp' directory
- Copy ZEIMA02.ZOWE.ZWEINIT.SZWESAMP(ZWESISTC) to /tmp/zwe-init-stc-3037
  * Succeeded
��/��ʀ�/�%������ǀ���>?���- cp  -v "/tmp/zwe-init-stc-3037" "//'ZEIMA02.ZOWE.ZWEINIT.JCLLIB(2)'"
  * Failed
ERROR:     cp: FSUM6259 target file "//'ZEIMA02.ZOWE.ZWEINIT.JCLLIB(2)'": EDC5061I An error occurred when attempting to define a file to the system. (errno2=0xC00B0403)
    /tmp/zwe-init-stc-3037 -> ZEIMA02.ZOWE.ZWEINIT.JCLLIB(2): text
ERROR: Error ZWEL0160E: Failed to write to ZEIMA02.ZOWE.ZWEINIT.JCLLIB(2). Please check if target data set is opened by others.
```

Already mentioned in some previous comment:
 - illegal name ZEIMA02.ZOWE.ZWEINIT.JCLLIB(2) was accepted and there was an attempt to save it
 - garbage in the command output

After changing STCs to valid names, there are problems, which are individually mentioned on the specific line.

Martin-Zeithaml commented on bin/commands/init/vsam/index.ts:
```diff
@@ -0,0 +1,166 @@
+      common.printTrace(stringlib.paddingLeft(theDataset.out, "    "));
+
+      const tmpFileContent = theDataset.out.replace(/\n\/\/\s*SET MODE=.*\n/, `\n//         SET  MODE=${vsam_mode}\n`)
+                .replace(new RegExp('\#dsname', 'g'), vsam_name)
```
This must be smarter, it is replacing text in the comments (which leads to trunc error)
From:
```
//*                                                                   
//* 2) Change all #dsname placeholders to the desired data set name   
//*    of the VSAM cluster. The maximum length is 38 characters.      
//*                                                                   
```
To:
```
//*                                                                             
//* 2) Change all ZEIMA02.ZOWE.ZWEINIT.CACHE placeholders to the desired data se
//*    of the VSAM cluster. The maximum length is 38 characters.                
//*
```

Martin-Zeithaml committed:
Small JCL manipulation changes

Signed-off-by: Martin Zeithaml <Martin.Zeithaml@broadcom.com>

Martin-Zeithaml committed:
Replace variables only in DEFINE CLUSTER

Signed-off-by: Martin Zeithaml <Martin.Zeithaml@broadcom.com>

Martin-Zeithaml commented on bin/commands/init/vsam/index.ts:
```diff
@@ -0,0 +1,166 @@
+      common.printTrace(stringlib.paddingLeft(theDataset.out, "    "));
+
+      const tmpFileContent = theDataset.out.replace(/\n\/\/\s*SET MODE=.*\n/, `\n//         SET  MODE=${vsam_mode}\n`)
+                .replace(new RegExp('[(]\#dsname', 'g'), '(' + vsam_name)
```
The shell version is replacing after the `//ALLOC` statement. This is simple way how to replace `#dsname`, `#volume` and `#storclas` in the `DEFINE CLUSTER` section only. It is searching for `(#variable` and replace it with `(value_of_the_variable`. It will keep `#variable` in the comment section unchanged. Feel free to modify it.

1000TurquoisePogs committed:
Merge branch 'v2.x/staging' into feature/v2/zwe-init-ts

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

struga0258 commented:
Pr is failing to build during pre-packaging.sh.. seems like manifest.json is missing 
```
cat: /ZOWE/tmp/pax-packaging-zowe-[168](https://github.com/zowe/zowe-install-packaging/actions/runs/5256332607/jobs/9497496582?pr=3135#step:13:169)6664364860/content/manifest.json: EDC5129I No such file or directory.
[pre-packaging.sh] Error: failed to extract version from manifest.json
```

Martin-Zeithaml committed:
Wrong variable handling

Signed-off-by: Martin Zeithaml <Martin.Zeithaml@broadcom.com>

Martin-Zeithaml commented on bin/libs/zos-dataset.ts:
```diff
@@ -235,7 +238,7 @@ export function getDatasetVolume(dataset: string): { rc: number, volume?: string
   const result = isDatasetSmsManaged(dataset);
   if (result.rc) {
-    common.printError("Error ZWEL0134E: Failed to find SMS status of data set ${dataset}.");
+    common.printError(`Error ZWEL0134E: Failed to find SMS status of data set ${dataset}.`);
```
This was printing `ERROR: Error ZWEL0134E: Failed to find SMS status of data set ${dataset}.` instead of `ERROR: Error ZWEL0134E: Failed to find SMS status of data set DATASET.USED.IN.ZOWE.YAML.`

Martin-Zeithaml committed:
Validate datasets using server-common

Signed-off-by: Martin Zeithaml <Martin.Zeithaml@broadcom.com>

Martin-Zeithaml committed:
Wrong text for message ZWEL0143E

Signed-off-by: Martin Zeithaml <Martin.Zeithaml@broadcom.com>

Martin-Zeithaml commented on schemas/zowe-yaml-schema.json:
## To do

- Add member validation for `zowe.setup.security.stcs.zowe + zis + aux`

Commit:
Merge branch 'v2.x/staging' of https://github.com/zowe/zowe-install-packaging into feature/v2/zwe-init-ts

struga0258 commented:
.pax/prepare-workspace.sh is throwing the following errors and not fully executing:
 
```
../../bin/libs/zos-jes.ts(86,8): error TS2339: Property 'sleep' does not exist on type 'typeof import("os")'.
../../bin/commands/init/mvs/index.ts(12,22): error TS2307: Cannot find module 'std' or its corresponding type declarations.
../../bin/commands/init/vsam/index.ts(12,22): error TS2307: Cannot find module 'std' or its corresponding type declarations.
../../bin/commands/init/vsam/index.ts(140,8): error TS2339: Property 'remove' does not exist on type 'typeof import("os")'.
../../bin/commands/init/apfauth/index.ts(12,22): error TS2307: Cannot find module 'std' or its corresponding type declarations.
../../bin/commands/init/security/index.ts(12,22): error TS2307: Cannot find module 'std' or its corresponding type declarations.
../../bin/commands/init/security/index.ts([104](https://github.com/zowe/zowe-install-packaging/actions/runs/5295831464/jobs/9586494827?pr=3135#step:13:105),6): error TS2339: Property 'remove' does not exist on type 'typeof import("os")'.
../../bin/commands/init/certificate/index.ts(12,22): error TS2307: Cannot find module 'std' or its corresponding type declarations.
../../bin/commands/init/stc/index.ts(13,22): error TS2307: Cannot find module 'std' or its corresponding type declarations.
../../bin/commands/init/stc/index.ts(168,8): error TS2339: Property 'remove' does not exist on type 'typeof import("os")'.
../../bin/commands/init/stc/index.ts(210,8): error TS2339: Property 'remove' does not exist on type 'typeof import("os")'.
../../bin/commands/init/stc/index.ts(249,8): error TS2339: Property 'remove' does not exist on type 'typeof import("os")'.
../../bin/commands/init/index.ts(12,22): error TS2307: Cannot find module 'std' or its corresponding type declarations.
../../bin/commands/init/mvs/cli.ts(12,22): error TS2307: Cannot find module 'std' or its corresponding type declarations.
../../bin/commands/init/security/cli.ts(12,22): error TS2307: Cannot find module 'std' or its corresponding type declarations.
../../bin/commands/init/stc/cli.ts(12,22): error TS2307: Cannot find module 'std' or its corresponding type declarations.
```

struga0258 commented:
This is where it fails: 
https://github.com/zowe/zowe-install-packaging/blob/v2.x/staging/.pax/prepare-workspace.sh#L122

Martin-Zeithaml committed:
Validate zowe.setup.dataset.security.stcs.* as zoweDatasetMember

Signed-off-by: Martin Zeithaml <Martin.Zeithaml@broadcom.com>

1000TurquoisePogs commented:
> .pax/prepare-workspace.sh is throwing the following errors and not fully executing:
> 
> ```
> ../../bin/libs/zos-jes.ts(86,8): error TS2339: Property 'sleep' does not exist on type 'typeof import("os")'.
> ../../bin/commands/init/mvs/index.ts(12,22): error TS2307: Cannot find module 'std' or its corresponding type declarations.
> ../../bin/commands/init/vsam/index.ts(12,22): error TS2307: Cannot find module 'std' or its corresponding type declarations.
> ../../bin/commands/init/vsam/index.ts(140,8): error TS2339: Property 'remove' does not exist on type 'typeof import("os")'.
> ../../bin/commands/init/apfauth/index.ts(12,22): error TS2307: Cannot find module 'std' or its corresponding type declarations.
> ../../bin/commands/init/security/index.ts(12,22): error TS2307: Cannot find module 'std' or its corresponding type declarations.
> ../../bin/commands/init/security/index.ts([104](https://github.com/zowe/zowe-install-packaging/actions/runs/5295831464/jobs/9586494827?pr=3135#step:13:105),6): error TS2339: Property 'remove' does not exist on type 'typeof import("os")'.
> ../../bin/commands/init/certificate/index.ts(12,22): error TS2307: Cannot find module 'std' or its corresponding type declarations.
> ../../bin/commands/init/stc/index.ts(13,22): error TS2307: Cannot find module 'std' or its corresponding type declarations.
> ../../bin/commands/init/stc/index.ts(168,8): error TS2339: Property 'remove' does not exist on type 'typeof import("os")'.
> ../../bin/commands/init/stc/index.ts(210,8): error TS2339: Property 'remove' does not exist on type 'typeof import("os")'.
> ../../bin/commands/init/stc/index.ts(249,8): error TS2339: Property 'remove' does not exist on type 'typeof import("os")'.
> ../../bin/commands/init/index.ts(12,22): error TS2307: Cannot find module 'std' or its corresponding type declarations.
> ../../bin/commands/init/mvs/cli.ts(12,22): error TS2307: Cannot find module 'std' or its corresponding type declarations.
> ../../bin/commands/init/security/cli.ts(12,22): error TS2307: Cannot find module 'std' or its corresponding type declarations.
> ../../bin/commands/init/stc/cli.ts(12,22): error TS2307: Cannot find module 'std' or its corresponding type declarations.
> ```

looks like this is caused by this branch not being up to date with the rest of staging.
os got renamed to cm_os, https://github.com/zowe/zowe-install-packaging/blob/v2.x/staging/bin/libs/common.ts#L13
so this branch needs to do the same.

struga0258 commented:
I merged staging into this pr before running that build, those errors are mostly related to the ts code.

1000TurquoisePogs commented:
Yes, point still stands though. I will fix it.

1000TurquoisePogs committed:
Switch os and std to cm_os and cm_std

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

struga0258 commented:
When testing `zwe init -c /path/to/zowe.yaml`

MVS: looks good 
VSAM: keeps throwing me an error that caching service is not configured to use VSAM even though it is in my yaml, this seems like there is a bug..

```
>> Create VSAM storage for Zowe Caching Service
ERROR: Warning ZWEL0301W: Zowe Caching Service is not configured to use VSAM. Command skipped.
-------------------------------------------------------------------------------
```

APF: looks good and works fine

SECURITY: it submitting the JCL but its throwing some weird symbols in the output..

Run Zowe security configurations

```
Modify ZWESECUR
▒▒/▒▒ʀ▒/▒%▒▒▒▒▒ǀ▒▒▒>?▒▒▒- TS3215.CUST.JCLLIB(ZW172076) is prepared
Submit TS3215.CUST.JCLLIB(ZW172076)
ERROR:   * Job (J0657100) doesn't finish within max waiting period.
- Job ZWESECUR(J0657100) ends with code 0 (COMPLETED).
```


CERTFICATE: So far i have only tried pck12 and its working fine..

STC: its working fine but also showing some weird symbols in the output...

>> Install Zowe main started task

```
Warning ZWEL0301W: ROCKET.USER.PROCLIB(ZWEMVDJS) already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Warning ZWEL0301W: ROCKET.USER.PROCLIB(ZWESISJS) already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Warning ZWEL0301W: ROCKET.USER.PROCLIB(ZWEMVD01) already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Modify ZWESLSTC and save as TS3215.CUST.JCLLIB(ZWEMVDJS)
▒▒/▒▒ʀ▒/▒%▒▒▒▒▒ǀ▒▒▒>?▒▒▒Modify ZWESISTC and save as TS3215.CUST.JCLLIB(ZWESISJS)
▒▒/▒▒ʀ▒/▒%▒▒▒▒▒ǀ▒▒▒>?▒▒▒Modify ZWESASTC and save as TS3215.CUST.JCLLIB(ZWEMVD01)
▒▒/▒▒ʀ▒/▒%▒▒▒▒▒ǀ▒▒▒>?▒▒▒
```

struga0258 commented:
When testing` zwe init -c /path/to/zowe.yaml  --allow-overwrite`

MVS: works fine if datasets aren't created, but when they are the allow-overwrite doesn't seem to work...
```
>> Initialize Zowe custom data sets

Create data sets if they do not exist
Warning ZWEL0301W: TS3215.USER.PARMLIB already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Warning ZWEL0301W: TS3215.CUST.JCLLIB already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Warning ZWEL0301W: TS3215.TESTX.ZWESALL already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Warning ZWEL0301W: TS3215.USER.SZWEPLUG already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Skipped writing to a dataset. To write, you must use --allow-overwrite.
```

VSAM: not working with following error 
`ERROR: Warning ZWEL0301W: Zowe Caching Service is not configured to use VSAM. Command skipped.`

APF: works fine
SECURITY: same issue as before symbols in log

CERTIFCATE: Again only used pck12 for this test and it worked fine, and allow-overwrite worked as well it overwrote the old one 

STC: I have added --allow-overwrite but i still get the following error:
```
>> Install Zowe main started task

Warning ZWEL0301W: TS3215.CUST.JCLLIB(ZWEMVDJS) already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Warning ZWEL0301W: ROCKET.USER.PROCLIB(ZWEMVDJS) already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Warning ZWEL0301W: TS3215.CUST.JCLLIB(ZWESISJS) already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Warning ZWEL0301W: ROCKET.USER.PROCLIB(ZWESISJS) already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Warning ZWEL0301W: TS3215.CUST.JCLLIB(ZWEMVD01) already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Warning ZWEL0301W: ROCKET.USER.PROCLIB(ZWEMVD01) already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Skipped writing to TS3215.CUST.JCLLIB. To write, you must use --allow-overwrite.
Skipped writing to ROCKET.USER.PROCLIB. To write, you must use --allow-overwrite.
```

struga0258 commented:
When testing `zwe init mvs` with all options. This command seems to be working as expected. allow-override works when used alone with zwe init mvs. 

```
bash-4.3$ zwe init mvs -c /proj/mvd/zowe-james/zowe/zowe.yaml
Temporary directory '/tmp/.zweenv-8112' created.
Zowe will remove it on success, but if zwe exits with a non-zero code manual cleanup would be needed.
-------------------------------------------------------------------------------
>> Initialize Zowe custom data sets

Create data sets if they do not exist
Creating TS3215.USER.PARMLIB
Creating TS3215.CUST.JCLLIB
Creating TS3215.TESTX.ZWESALL
Creating TS3215.USER.SZWEPLUG
Copy TS3215.TESTX.SZWESAMP(ZWESIP00) to TS3215.USER.PARMLIB(ZWESIP00)
Copy components/zss/LOADLIB/ZWESIS01 to TS3215.TESTX.ZWESALL(ZWESIS01)
Copy components/zss/LOADLIB/ZWESAUX to TS3215.TESTX.ZWESALL(ZWESAUX)
Copy components/launcher/bin/zowe_launcher to TS3215.TESTX.ZWESALL(ZWELNCH)

>> Zowe custom data sets are initialized successfully.

Temporary directory /tmp/.zweenv-8112 removed successfully.
bash-4.3$ zwe init mvs -c /proj/mvd/zowe-james/zowe/zowe.yaml
Temporary directory '/tmp/.zweenv-8012' created.
Zowe will remove it on success, but if zwe exits with a non-zero code manual cleanup would be needed.
-------------------------------------------------------------------------------
>> Initialize Zowe custom data sets

Create data sets if they do not exist
Warning ZWEL0301W: TS3215.USER.PARMLIB already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Warning ZWEL0301W: TS3215.CUST.JCLLIB already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Warning ZWEL0301W: TS3215.TESTX.ZWESALL already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Warning ZWEL0301W: TS3215.USER.SZWEPLUG already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Skipped writing to a dataset. To write, you must use --allow-overwrite.

>> Zowe custom data sets are initialized successfully.

Temporary directory /tmp/.zweenv-8012 removed successfully.
bash-4.3$ zwe init mvs -c /proj/mvd/zowe-james/zowe/zowe.yaml --allow-overwrite
Temporary directory '/tmp/.zweenv-4766' created.
Zowe will remove it on success, but if zwe exits with a non-zero code manual cleanup would be needed.
-------------------------------------------------------------------------------
>> Initialize Zowe custom data sets

Create data sets if they do not exist
Warning ZWEL0300W: TS3215.USER.PARMLIB already exists. Members in this data set will be overwritten.
Warning ZWEL0300W: TS3215.CUST.JCLLIB already exists. Members in this data set will be overwritten.
Warning ZWEL0300W: TS3215.TESTX.ZWESALL already exists. Members in this data set will be overwritten.
Warning ZWEL0300W: TS3215.USER.SZWEPLUG already exists. Members in this data set will be overwritten.
Copy TS3215.TESTX.SZWESAMP(ZWESIP00) to TS3215.USER.PARMLIB(ZWESIP00)
Copy components/zss/LOADLIB/ZWESIS01 to TS3215.TESTX.ZWESALL(ZWESIS01)
Copy components/zss/LOADLIB/ZWESAUX to TS3215.TESTX.ZWESALL(ZWESAUX)
Copy components/launcher/bin/zowe_launcher to TS3215.TESTX.ZWESALL(ZWELNCH)

>> Zowe custom data sets are initialized successfully.

Temporary directory /tmp/.zweenv-4766 removed successfully.
```

struga0258 commented:
When testing `zwe init apf`, everything works as expected. 

```
bash-4.3$ zwe init apfauth -c /proj/mvd/zowe-james/zowe/zowe.yaml
Temporary directory '/tmp/.zweenv-2399' created.
Zowe will remove it on success, but if zwe exits with a non-zero code manual cleanup would be needed.
-------------------------------------------------------------------------------
>> APF authorize load libraries

APF authorize TS3215.TESTX.ZWESALL
APF authorize TS3215.USER.SZWEPLUG

>> Zowe load libraries are APF authorized successfully.
```

struga0258 commented:
When testing `zwe init vsam`, seems to not be working, my yaml:

```
zowe:
  setup:
    dataset:
      prefix: TS3215.TESTX
      jcllib: TS3215.CUST.JCLLIB
    vsam:
      mode: NONRLS
      volume: VOL123
      storageClass: ''
components:
  caching-service:
    storage:
      mode: VSAM
      vsam:
        name: TS3215.CACHE
```

The error:
```
bash-4.3$ zwe init vsam  -c /proj/mvd/zowe-james/zowe/zowe.yaml --allow-overwrite
Temporary directory '/tmp/.zweenv-7877' created.
Zowe will remove it on success, but if zwe exits with a non-zero code manual cleanup would be needed.
-------------------------------------------------------------------------------
>> Create VSAM storage for Zowe Caching Service

ERROR: Warning ZWEL0301W: Zowe Caching Service is not configured to use VSAM. Command skipped.
Temporary directory /tmp/.zweenv-7877 removed successfully.
```

struga0258 commented:
When testing `zwe init stc `, only issue found was when i add --allow-overwrite it throws some weird symbols in the output. 

```
bash-4.3$ zwe init stc  -c /proj/mvd/zowe-james/zowe/zowe.yaml
Temporary directory '/tmp/.zweenv-4284' created.
Zowe will remove it on success, but if zwe exits with a non-zero code manual cleanup would be needed.
-------------------------------------------------------------------------------
>> Install Zowe main started task

Warning ZWEL0301W: TS3215.CUST.JCLLIB(ZWEMVDJS) already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Warning ZWEL0301W: TS3215.CUST.JCLLIB(ZWESISJS) already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Warning ZWEL0301W: TS3215.CUST.JCLLIB(ZWEMVD01) already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Skipped writing to TS3215.CUST.JCLLIB. To write, you must use --allow-overwrite.
Copy TS3215.CUST.JCLLIB(ZWEMVDJS) to ROCKET.USER.PROCLIB(ZWEMVDJS)
Copy TS3215.CUST.JCLLIB(ZWESISJS) to ROCKET.USER.PROCLIB(ZWESISJS)
Copy TS3215.CUST.JCLLIB(ZWEMVD01) to ROCKET.USER.PROCLIB(ZWEMVD01)

>> Zowe main started tasks are installed successfully.

Temporary directory /tmp/.zweenv-4284 removed successfully.
bash-4.3$ zwe init stc  -c /proj/mvd/zowe-james/zowe/zowe.yaml
Temporary directory '/tmp/.zweenv-4639' created.
Zowe will remove it on success, but if zwe exits with a non-zero code manual cleanup would be needed.
-------------------------------------------------------------------------------
>> Install Zowe main started task

Warning ZWEL0301W: TS3215.CUST.JCLLIB(ZWEMVDJS) already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Warning ZWEL0301W: ROCKET.USER.PROCLIB(ZWEMVDJS) already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Warning ZWEL0301W: TS3215.CUST.JCLLIB(ZWESISJS) already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Warning ZWEL0301W: ROCKET.USER.PROCLIB(ZWESISJS) already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Warning ZWEL0301W: TS3215.CUST.JCLLIB(ZWEMVD01) already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Warning ZWEL0301W: ROCKET.USER.PROCLIB(ZWEMVD01) already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Skipped writing to TS3215.CUST.JCLLIB. To write, you must use --allow-overwrite.
Skipped writing to ROCKET.USER.PROCLIB. To write, you must use --allow-overwrite.

>> Zowe main started tasks are installed successfully.

Temporary directory /tmp/.zweenv-4639 removed successfully.

bash-4.3$ zwe init stc  -c /proj/mvd/zowe-james/zowe/zowe.yaml --allow-overwrite
Temporary directory '/tmp/.zweenv-2782' created.
Zowe will remove it on success, but if zwe exits with a non-zero code manual cleanup would be needed.
-------------------------------------------------------------------------------
>> Install Zowe main started task

Warning ZWEL0300W: TS3215.CUST.JCLLIB(ZWEMVDJS) already exists. This data set member will be overwritten during configuration.
Warning ZWEL0300W: ROCKET.USER.PROCLIB(ZWEMVDJS) already exists. This data set member will be overwritten during configuration.
Warning ZWEL0300W: TS3215.CUST.JCLLIB(ZWESISJS) already exists. This data set member will be overwritten during configuration.
Warning ZWEL0300W: ROCKET.USER.PROCLIB(ZWESISJS) already exists. This data set member will be overwritten during configuration.
Warning ZWEL0300W: TS3215.CUST.JCLLIB(ZWEMVD01) already exists. This data set member will be overwritten during configuration.
Warning ZWEL0300W: ROCKET.USER.PROCLIB(ZWEMVD01) already exists. This data set member will be overwritten during configuration.
Modify ZWESLSTC and save as TS3215.CUST.JCLLIB(ZWEMVDJS)
▒▒/▒▒ʀ▒/▒%▒▒▒▒▒ǀ▒▒▒>?▒▒▒Modify ZWESISTC and save as TS3215.CUST.JCLLIB(ZWESISJS)
▒▒/▒▒ʀ▒/▒%▒▒▒▒▒ǀ▒▒▒>?▒▒▒Modify ZWESASTC and save as TS3215.CUST.JCLLIB(ZWEMVD01)
▒▒/▒▒ʀ▒/▒%▒▒▒▒▒ǀ▒▒▒>?▒▒▒
Copy TS3215.CUST.JCLLIB(ZWEMVDJS) to ROCKET.USER.PROCLIB(ZWEMVDJS)
Copy TS3215.CUST.JCLLIB(ZWESISJS) to ROCKET.USER.PROCLIB(ZWESISJS)
Copy TS3215.CUST.JCLLIB(ZWEMVD01) to ROCKET.USER.PROCLIB(ZWEMVD01)

>> Zowe main started tasks are installed successfully.

Temporary directory /tmp/.zweenv-2782 removed successfully.
```

struga0258 commented:
when testing `zwe init --skip-security-setup`, everything works as expected. skips apf and security.

Martin-Zeithaml commented:
The garbage for some commands is caused by `changeTag` and the garbage itself is a piece of code in binary, it starts with CLC, branch, ... and ends with SVC 0:
```
0000000000      D5D7    5DD5    D75D    07D5    D75D    D5D7    5DAD    D8D5
0000000020      D75D    07D5    D75D    85D5    D75D    D5D7    5DD5    D75D
0000000040      D5D7    5DD5    D75D    D5D7    5D47    D8D5    D75D    D5D7
0000000060      5DD5    D75D    9E1A    0A00
```
This one is from `zwe init security`.

But why is the code printed and not executed is a mystery to me.

**Edit:**
`bin/libs/zos-fs.ts`
Locally changed in `ensureFileEncoding` from `zos.changeTag(file, 0);` to ``shell.execSync('sh', '-c', `chtag -r ${file}`);`` and it is working fine.

JoeNemo commented:
So, Is this all good now, and just needs a review??

Martin-Zeithaml committed:
changeTag replaced with shell version

Signed-off-by: Martin Zeithaml <Martin.Zeithaml@broadcom.com>

Martin-Zeithaml commented:
I suggest following:

- Mandatory step
  - Review new [solution](https://github.com/zowe/zowe-install-packaging/pull/3135/files#diff-626d8a6b406d394c8d2eb7ed96f92e9f620afdd8094f51c3039e2dfc5bdc43d8)
  - There was already one `changeTag` commented out, is this correct?
- Optional, but strongly recommended step, review what is happening:
  - What are we printing? Is it `changeTag` code? Or worse, are we printing some random piece of memory?
  - Why the heck are we printing a piece of memory?

Martin-Zeithaml commented:
Another "family" of bugs: `--silent` is either not working or only partially, see examples:
## /git/qa.zweInit/apfauth: zwe init apfauth -c auth.cm.ok.yaml
```
Temporary directory '/tmp/.zweenv-4658' created.
Zowe will remove it on success, but if zwe exits with a non-zero code manual cleanup would be needed.
-------------------------------------------------------------------------------
>> APF authorize load libraries

APF authorize ZOWE.ZWEINIT.TEST1
APF authorize ZOWE.ZWEINIT.TEST2

>> Zowe load libraries are APF authorized successfully.

Temporary directory /tmp/.zweenv-4658 removed successfully.
```
## /git/qa.zweInit/apfauth: zwe init apfauth -c auth.cm.ok.yaml --silent
```
Temporary directory '/tmp/.zweenv-6730' created.
Zowe will remove it on success, but if zwe exits with a non-zero code manual cleanup would be needed.
-------------------------------------------------------------------------------
>> APF authorize load libraries

APF authorize ZOWE.ZWEINIT.TEST1
APF authorize ZOWE.ZWEINIT.TEST2

>> Zowe load libraries are APF authorized successfully.

Temporary directory /tmp/.zweenv-6730 removed successfully.
```
## /git/qa.zweInit/stc: zwe init stc -c stc.cm.ok.yaml
```
paths=["/","/tmp","/tmp/.zweenv-1769"]
firstMissingDir=/tmp/.zweenv-1769
Temporary directory '/tmp/.zweenv-1769' created.
Zowe will remove it on success, but if zwe exits with a non-zero code manual cleanup would be needed.
-------------------------------------------------------------------------------
>> Install Zowe main started task

Warning ZWEL0301W: ZOWE.ZWEINIT.JCLLIB(MZZOWE) already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Warning ZWEL0301W: ZOWE.ZWEINIT.PROCLIB(MZZOWE) already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Warning ZWEL0301W: ZOWE.ZWEINIT.JCLLIB(MZZIS) already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Warning ZWEL0301W: ZOWE.ZWEINIT.PROCLIB(MZZIS) already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Warning ZWEL0301W: ZOWE.ZWEINIT.JCLLIB(MZAUX) already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Warning ZWEL0301W: ZOWE.ZWEINIT.PROCLIB(MZAUX) already exists and will not be overwritten. For upgrades, you must use --allow-overwrite.
Skipped writing to ZOWE.ZWEINIT.JCLLIB. To write, you must use --allow-overwrite.
Skipped writing to ZOWE.ZWEINIT.PROCLIB. To write, you must use --allow-overwrite.

>> Zowe main started tasks are installed successfully.

Temporary directory /tmp/.zweenv-1769 removed successfully.
```
## /git/qa.zweInit/stc: zwe init stc -c stc.cm.ok.yaml --silent
```
Temporary directory '/tmp/.zweenv-6238' created.
Zowe will remove it on success, but if zwe exits with a non-zero code manual cleanup would be needed.
Temporary directory /tmp/.zweenv-6238 removed successfully.
```

Martin-Zeithaml committed:
Missing parameters

Signed-off-by: Martin Zeithaml <Martin.Zeithaml@broadcom.com>

Martin-Zeithaml committed:
Init stc is able to handle long DSN and split line if needed

Signed-off-by: Martin Zeithaml <Martin.Zeithaml@broadcom.com>

1000TurquoisePogs commented:
about silent/not, every command that's not under the "internal" folder does temp dir stuff that if fails would need cleanup, so we warn about that.
it could be made silent on the "silent" flag, especially since the silent flag is an env var that can be detected in that code.

Martin-Zeithaml committed:
Update init certificate 

Missing slash in warning ZWEL0300

Signed-off-by: Martin Zeithaml <66114686+Martin-Zeithaml@users.noreply.github.com>

Martin-Zeithaml committed:
Update init certificate

Missing slash in ZWEL0300W

Signed-off-by: Martin Zeithaml <66114686+Martin-Zeithaml@users.noreply.github.com>

Martin-Zeithaml committed:
toLowerCase for boolean

This does not work under `node` or `configmgr`:
```
let x = true;
console.log(x.toLowerCase());    
```
The boolean must be converted first before lowercased.

Signed-off-by: Martin Zeithaml <66114686+Martin-Zeithaml@users.noreply.github.com>

Martin-Zeithaml committed:
JCERACFKS -> JCE*KS

Signed-off-by: Martin Zeithaml <Martin.Zeithaml@broadcom.com>

Martin-Zeithaml committed:
Update ZWEL0164E for init certificate

Signed-off-by: Martin Zeithaml <66114686+Martin-Zeithaml@users.noreply.github.com>

Martin-Zeithaml committed:
Typo in variable name

pkcs12_diryectory -> pkcs12_directory

Signed-off-by: Martin Zeithaml <66114686+Martin-Zeithaml@users.noreply.github.com>

Martin-Zeithaml committed:
zwe init vsam: dataset operations change

Signed-off-by: Martin Zeithaml <Martin.Zeithaml@broadcom.com>

Martin-Zeithaml committed:
Merge branch 'v2.x/staging' into feature/v2/zwe-init-ts

Signed-off-by: Martin Zeithaml <66114686+Martin-Zeithaml@users.noreply.github.com>

1000TurquoisePogs committed:
Update index.ts for acting upon ZWESISDL

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

1000TurquoisePogs committed:
Merge branch 'v2.x/staging' into feature/v2/zwe-init-ts

1000TurquoisePogs committed:
Work-around init problem for detecting java and node homes when user did not specify them

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

1000TurquoisePogs committed:
Update main.yml to only apply java and node homes when value not null

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

1000TurquoisePogs committed:
fixing missing quotes

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

1000TurquoisePogs committed:
remove unneeded pem section

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

1000TurquoisePogs committed:
Fix ca name, fix zosmf trust, and shorten retry time to not waste build time on failures

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

struga0258 commented:
Generating a keyring using Zowe seems to not be working, I get the following error on multiple systems:
```
bash-4.3$ zwe init certificate -c /proj/mvd/zowe-james/zowe/zowe.yaml --update-config
Temporary directory '/tmp/.zweenv-6545' created.
Zowe will remove it on success, but if zwe exits with a non-zero code manual cleanup would be needed.
bos extend currSize=0x0 dataSize=0x100c chunk=0x1000 extend=0x100c
-------------------------------------------------------------------------------
>> Initializing Zowe keystore

TypeError: cannot read property 'toLowerCase' of undefined
    at execute (/proj/mvd/zowe-james/zowe/bin/commands/init/certificate/index:181)
    at <anonymous> (/proj/mvd/zowe-james/zowe/bin/commands/init/certificate/cli.js:13)
```
Here is what i have in my yaml:
```
zowe:
  setup:
    certificate:
      keyring:
        name: ZoweKeyring
        label: localhost
        caLabel: localca
      type: JCERACFKS
```

Martin-Zeithaml committed:
Update PKCS12 values only if defined

Signed-off-by: Martin Zeithaml <Martin.Zeithaml@broadcom.com>

Martin-Zeithaml commented:
This [commit](https://github.com/zowe/zowe-install-packaging/pull/3135/commits/ee6edfdc741b4ce927a0533c14d563f46e331836) should solve it.

1000TurquoisePogs committed:
Merge branch 'v2.x/staging' into feature/v2/zwe-init-ts

1000TurquoisePogs committed:
Handle DETECT case for node and java

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

1000TurquoisePogs committed:
Wrap call to readyaml with requirenode

Signed-off-by: 1000TurquoisePogs <sgrady@rocketsoftware.com>

Martin-Zeithaml commented:
`certificate.sh` (used by `zwe init certificate`) is working for fine for "good" input, but... not for my crazy testing scenarios 😄 
No doubt, following examples are unrealistic, but legal (there is no schema limitation):
```
zowe.setup.certificate.keyring.name = "'"
zowe.setup.certificate.keyring.name = "AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA"
zowe.setup.certificate.keyring.name = "&SYSUID"
zowe.setup.certificate.keyring.name = "USER/MYRING"
```

The result is:
* Unescaped  single apostrophe breaks the JCL strings
* Too long name causes the write error
* `&` confuse `sed` and produces `//         SET  ZOWERING='//         SET ZOWERING='ZoweKeyring'SYSUID' `
* Slash confuse `sed` more and result is empty JCL

There is no useful schema validation for key ring name (doc I have found):
* ACF2
  * 237 characters
  * Mixed-cased characters and special characters: `(!&*¬-%_?:=.)`
  * Forbidden: space
* TSS
  * 237 characters
* RACF
  * 237 characters
  * Forbidden: `&*%`
Based on this we can only limit keyring name to 237 chars, otherwise we need to consider any char as valid.

Similar situation for the rest of the fields used in `certificate.sh`, like `CN, OU, O, L, ...`

**Conclusion**
* It is working fine for "normal" input
* I am **not** volunteer for fixing this in shell 😄 
* Let's do it better (some day) in Typescript

Martin-Zeithaml committed:
certificate keyring-jcl clean only for option 1

Signed-off-by: Martin Zeithaml <martin.zeithaml@broadcom.com>

Martin-Zeithaml committed:
Revert "certificate keyring-jcl clean only for option 1"

This reverts commit 8f296b593358638dc1523fe4ee20b1da5d59c0b6.

Martin-Zeithaml committed:
Tracing and message with details

Signed-off-by: Martin Zeithaml <Martin.Zeithaml@broadcom.com>

Martin-Zeithaml commented:
There is a related problem to [3140](https://github.com/zowe/zowe-install-packaging/issues/3140):
```
/zowe/bin/zwe init certificate scenario4.yaml --allow-overwrite
Warning ZWEL0300W: Keyring "safkeyring:////ZWESVUSR/TESTZoweKeyring1234" will be overwritten during configuration.
Error ZWEL0106E: ca-alias parameter is required
Try --help to get information about how to use this command.
```
* User gets the error message:
  * `Error ZWEL0106E: ca-alias parameter is required`
* And suggestion to see help:
  * `Try --help to get information about how to use this command.`
* Command executed by user is `zwe init certificate`, but the failing command is `zwe certificate keyring-jcl clean`
* User will probably try help for `zwe init certificate`
  * `zwe init certificate --help | grep -i 'ca-alias'` -> nothing

This [commit](https://github.com/zowe/zowe-install-packaging/pull/3135/commits/ddf75805e458e9045b10ef341d96f5953764a521) enhances:
* the tracing to include details about executed command
* error message
#### Shell tracing
```
- zwecli_inline_execute_command
  * certificate keyring-jcl clean --dataset-prefix ZOWE.ZWEINIT --jcllib ZOWE.ZWEINIT.JCLLIB --keyring-owner ZWESVUSR --keyring-name TESTZoweKeyring1234 --alias localhost --ca-alias  --security-product TSS
Error ZWEL0106E: ca-alias parameter is required
```
#### JavaScript tracing
```
- execZweSync
  * certificate keyring-jcl clean --dataset-prefix "ZOWE.ZWEINIT" --jcllib "ZOWE.ZWEINIT.JCLLIB" --keyring-owner "ZWESVUSR" --keyring-name "TESTZoweKeyring1234" --alias "localhost" --ca-alias "" --security-product "TSS"
Error ZWEL0106E: ca-alias parameter is required
```
#### JavaScript error message
  * Before: `ERROR: Error ZWEL0305E: Failed to call certificate command, rc=106.`
  * Now: `ERROR: Error ZWEL0305E: Failed to call "certificate keyring-jcl clean" command, rc=106.`
